library/src/main/java/com/bumptech/glide/load/resource/bitmap/ImageReader.java

```diff
@@ -245,8 +341,199 @@ public int getImageOrientation() throws IOException {
     }
 
     @Override
-    public void stopGrowingBuffers() {
-      // Nothing to do here.
+    public void stopGrowingBuffers() {}
+
+    /**
+     * Returns a decoded bitmap for the input stream, ensuring that any associated gainmap is
+     * decoded without errors on Android U if it is a valid gainamp.
+     *
+     * <p>Callers should use the method to decode streams with {@link
+     * BitmapFactory#decodeFileDescriptor(FileDescriptor, Rect, Options)} on Android U.
+     *
+     * @param parcelFileDescriptor from which the bitmap will be decoded.
+     * @param options to be applied in the {@link BitmapFactory#decodeStream} call.
+     */
+    @RequiresApi(VERSION_CODES.UPSIDE_DOWN_CAKE)
+    @Nullable
+    private static Bitmap safeDecodeHardwareBitmapWithGainmap(
+        ParcelFileDescriptor parcelFileDescriptor, Options options) {
+      Bitmap softwareBitmap = null;
+      try {
+        options.inPreferredConfig = Bitmap.Config.ARGB_8888;
+        softwareBitmap =
+            BitmapFactory.decodeFileDescriptor(
+                parcelFileDescriptor.getFileDescriptor(), /* outPadding= */ null, options);
+        if (softwareBitmap == null) {
+          return null;
+        }
+        return AndroidUHardwareBitmapFactoryDecoder.safeDecodeBitmapWithGainmap(softwareBitmap);
+      } finally {
+        if (softwareBitmap != null) {
+          softwareBitmap.recycle();
+        }
+        options.inPreferredConfig = Config.HARDWARE;
```

Comment:
nit: It took me a second to convince myself that you're just resetting it to what it was before, but I think it would be easier to reason about this method if you add an assertion at the start of it that the options are indeed set to `HARDWARE`.

Reply:
Done.